### ISSUE TITLE
Programatic dragging

### ISSUE THREAD

**Issue by alexreardon** [atlassian/react-beautiful-dnd#162]
Add the ability to control an entire drag programatically. This opens up a large number of experiences such as:

- voice controlled dragging
- virtual reality inputs
- animated rollbacks of actions

**Comment by theneva** [comment]
I'm trying to write an integration test with Cypress that uses events (`mousedown`, `mousemove`, and `mouseup`) to simulate a drag to assert that my form behaves correctly.

I've been completely unable to even get the draggable element to move in the test (and also in more "manual" tests of just firing events from the console in Chrome).

Is what I'm trying to do impossible without this idea implemented?

**Comment by CarrierDirectRoss** [comment]
I am unable to programmatically trigger DnD experiences too (in my tests). Are any examples available?

**Comment by CarrierDirectRoss** [comment]
Thanks for the response @alexreardon. The example definitely helped get me started. My current problem though is that I am able to programmatically initiate a drag on a DragHandle (via spacebar), but cannot move the item using the proper arrow keys for keyboard navigation. Below is a code snippet if you can spot anything wrong (using [react-testing-library](https://github.com/kentcdodds/react-testing-library))

```
const wrapper = render(<Yard />)
const track1El = wrapper.getByTestId(`droppable-${tracks.t1.id}`)
const beforeDragCars = track1El.querySelectorAll(
  '[data-testid^=draggable-]',
)
const beforeDragFirstText = beforeDragCars[0].textContent
const beforeDragSecondText = beforeDragCars[1].textContent

const dragHandle = getDragHandle(wrapper.container, cars.c1.id)

// initiate drag - This works!
Simulate.keyDown(dragHandle, {
  keyCode: keyCodes.space, // 32
})

const dragCountEl = await waitForElement(() =>
  wrapper.getByTestId('drag-count'),
)
expect(dragCountEl).toHaveTextContent('1')

// move right 1 position in list - This does NOT work
Simulate.keyDown(dragHandle, {
  keyCode: keyCodes.arrowRight, // 39
})

// drop item
Simulate.keyDown(dragHandle, {
  keyCode: keyCodes.space,
})
expect(wrapper.queryByTestId('drag-count')).toBeNull()

track1El = wrapper.getByTestId(`droppable-${tracks.t1.id}`)
const afterDragCars = track1El.querySelectorAll(
  '[data-testid^=draggable-]',
)
const afterDragFirstText = afterDragCars[0].textContent
const afterDragSecondText = afterDragCars[1].textContent

// assert that first & second items have swapped
expect(afterDragSecondText).toBe(beforeDragFirstText)
expect(afterDragFirstText).toBe(beforeDragSecondText)
```

**Comment by gyfchong** [comment]
Would this cover creating our own custom controls that could trigger a DnD experience? ie. up and down arrows which move items up an down a list when clicked.

**Comment by alexreardon** [comment]
Yes

**Comment by tim-soft** [comment]
This seems like a really cool idea. I'm imagining an API like a chess move i.e. "move a4 to c6"

https://en.m.wikipedia.org/wiki/Algebraic_notation_(chess)

There are really two ideal actions in my mind. You could have the foreshadowing of a drag where the droppable space opens up in the destination list, then commit to the actual drag and the droppable flies into place.

**Comment by alexreardon** [comment]
Because we have a state driven drag model there are a lot of possibilities

**Comment by millarm** [comment]
OK, we hit this same issue this week - we have some complicated business logic that runs when a drop happens, and we wanted to be able to test this programatically with cypress.io

Thumbs up to the idea of providing specific events that enable the scripting of this (in fact, from review yesterday, it looks like this would be reasonably easy - create a new sensor - in drag-handle/sensor that is a "SyntheticEventSensor" that responds to to synthetic events for "dragstart"/"dragmove"/"dragend" which would allow any JS test framework (or application code) that can trigger JS events to simulate dragging, in an explicit way without having to worry about the implementation of the mouse/touch/keyboard interfaces.

Anyway -> for anyone, like me hitting this problem, here's a recipe to make it work - this is now running in our product (test!) environment:

Using cypress.io to simulate a drag event do the following:

(our application is a calendar so we freeze the time of the tests in cypress - hence the use of cy.clock)

```
    const now = new Date(2018, 4, 11, 9, 0, 0).getTime(); // Freeze the time for the tests
    cy.clock(now);
    cy.visit('/');
    cy
    cy
      .get('.draggable-item-id-from-your-system-here') // or use [data-cy=test-id] for more robust tests
      .trigger('mousedown', { button: 0, clientX: 633, clientY: 440 }) // drag events test for button: 0 and also use the clientX and clientY values - the clientX and clientY values will be specific to your system
      .trigger('mousemove', { button: 0, clientX: 639, clientY: 500 }) // We perform a small move event of > 5 pixels this means we don't get dismissed by the sloppy click detection
      .tick(200); // react-beautiful-dnd has a minimum 150ms timeout before starting a drag operation, so wait at least this long.
    cy
      .get('.top-level-view') // now we perform drags on the whole screen, not just the draggable
      .trigger('mousemove', { button: 0, clientX: 939, clientY: 900 })
      .tick(1); // We've frozen time, so we need at least a 1ms tick to move to lead to the animations being executed. You should see the actual animation in the cypress window
    cy.get('.top-level-view').trigger('mouseup'); // Causes the drop to be run
    // Can now test the application's post DROP state
```

**Comment by CarrierDirectRoss** [comment]
Thank you @millarm. We have been using keyboard events to trigger drags in cypress tests. Now we can finally test all the use cases!

**Comment by alexreardon** [comment]
Our browser tests are now executing keyboard, mouse and touch dragging in chrome and firefox with test cafe. Feel free to take a look: https://github.com/atlassian/react-beautiful-dnd/blob/master/test/browser/simple-list.js

**Comment by alexreardon** [comment]
Relevant: #623

**Comment by KeevanDance** [comment]
Question about dragging while testing... 

I am using CRA, and React-Testing-Library fyi

In this example, I can successfully drag and drop an element from the top of a list, to the bottom of the list.
```
fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'Tab',
    keyCode: 9,
    which: 9,
  });

  await wait()

  // click spacebar to grab the item
  fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'Space',
    keyCode: 32,
    which: 32,
  });

  await wait()

  // down arrow to move lower in the list
  fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'ArrowDown',
    keyCode: 40,
    which: 40,
  });

  await wait();

  // space bar to drop the item
  fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'Space',
    keyCode: 32,
    which: 32,
  });

await wait();
```

However, I cannot get the item to from from a list into a different list right next to it. (the below code doesn't move the item)
```
fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'Tab',
    keyCode: 9,
    which: 9,
  });

  await wait()

  // click spacebar to grab the item
  fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'Space',
    keyCode: 32,
    which: 32,
  });

  await wait()

  // down arrow to move lower in the list
  fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'ArrowRight',
    keyCode: 39,
    which: 39,
  });

  await wait();

  // space bar to drop the item
  fireEvent.keyDown(getByTestId('draggable__uniqueId'), {
    key: 'Space',
    keyCode: 32,
    which: 32,
  });

await wait();
```

**Comment by maku-zuhlke** [comment]
> OK, we hit this same issue this week - we have some complicated business logic that runs when a drop happens, and we wanted to be able to test this programatically with cypress.io
> 
> Thumbs up to the idea of providing specific events that enable the scripting of this (in fact, from review yesterday, it looks like this would be reasonably easy - create a new sensor - in drag-handle/sensor that is a "SyntheticEventSensor" that responds to to synthetic events for "dragstart"/"dragmove"/"dragend" which would allow any JS test framework (or application code) that can trigger JS events to simulate dragging, in an explicit way without having to worry about the implementation of the mouse/touch/keyboard interfaces.
> 
> Anyway -> for anyone, like me hitting this problem, here's a recipe to make it work - this is now running in our product (test!) environment:
> 
> Using cypress.io to simulate a drag event do the following:
> 
> (our application is a calendar so we freeze the time of the tests in cypress - hence the use of cy.clock)
> 
> ```
>     const now = new Date(2018, 4, 11, 9, 0, 0).getTime(); // Freeze the time for the tests
>     cy.clock(now);
>     cy.visit('/');
>     cy
>     cy
>       .get('.draggable-item-id-from-your-system-here') // or use [data-cy=test-id] for more robust tests
>       .trigger('mousedown', { button: 0, clientX: 633, clientY: 440 }) // drag events test for button: 0 and also use the clientX and clientY values - the clientX and clientY values will be specific to your system
>       .trigger('mousemove', { button: 0, clientX: 639, clientY: 500 }) // We perform a small move event of > 5 pixels this means we don't get dismissed by the sloppy click detection
>       .tick(200); // react-beautiful-dnd has a minimum 150ms timeout before starting a drag operation, so wait at least this long.
>     cy
>       .get('.top-level-view') // now we perform drags on the whole screen, not just the draggable
>       .trigger('mousemove', { button: 0, clientX: 939, clientY: 900 })
>       .tick(1); // We've frozen time, so we need at least a 1ms tick to move to lead to the animations being executed. You should see the actual animation in the cypress window
>     cy.get('.top-level-view').trigger('mouseup'); // Causes the drop to be run
>     // Can now test the application's post DROP state
> ```

Thanks this worked beautifully for me!

**Comment by abzainuddin** [comment]
Tweaked @goncy solution a bit, so it works on elements instead of using offsets (tested on version 10.0.3):

`commands/dragAndDrop.js`

```
const sloppyClickThreshold = 5;
const primaryButton = 0;

export function drag(subject) {
    const coords = subject[0].getBoundingClientRect();

    // Hoops we need to jump through to register a mouse drag.
    cy.wrap(subject)
        .trigger('mousedown', {
            // Hoop for checking primary button.
            button: primaryButton,

            // Register clientX + clientY for sloppy click detection in
            // subsequent mousemove.
            clientX: coords.left,
            clientY: coords.top
        })
        .trigger('mousemove', {
            button: primaryButton,

            // Make sure we pass sloppyClickThreshold detection.
            clientX: coords.left + sloppyClickThreshold,
            clientY: coords.top
        });
};

export function drop(subject) {
    cy.wrap(subject)
        .trigger('mousemove', { button: primaryButton })
        .trigger('mouseup');
};
```

`commands.js`

```
import { drag, drop } from './dragAndDrop';

Cypress.Commands.add('drag', { prevSubject: 'element' }, drag);
Cypress.Commands.add('drop', { prevSubject: 'element' }, drop);
```
Usage:
```
cy.get('.draggable').drag();
// Do some tests here, for example check change in droppable background colors.
cy.get('.droppable').drop();
```

**Comment by tcoughlin3** [comment]
@goncy's version of @millarm's command is working great for me in the Cypress GUI but fails to move anything in the Cypress Electron browser. Any suggestions are greatly appreciated.

https://stackoverflow.com/questions/54120227/drag-and-drop-cypress-test-fails-in-electron-browser-only

**Comment by nmain** [comment]
One potential use case I have that could be covered by this feature:  I have a draggable list, and some button outside of that list that sorts it in a particular way.  When I click the sort button, I'd like the list to sort, _and animate as it sorts_.  If there was programatic dragging, my sort code could just fire off a series of drags.

There are other libraries that do this, but if I'm already using `react-beautiful-dnd` and my users are already used to exactly its animations when they drag, why not kill every bird with one giant boulder?

**Comment by alexreardon** [comment]
It looks like this might be going out as a part of #1225

**Comment by alexreardon** [comment]
![programatic-dragging 2019-05-07 21_44_54](https://user-images.githubusercontent.com/2182637/57420906-f5e4bc00-724c-11e9-805c-afadfdefb5c4.gif)

**Comment by alexreardon** [comment]
### Ideas for programmatic API examples

- Voice `sensor`. Lower level control for voice dragging. It would be it's own `npm` package /cc @danieldelcore
- Higher lever voice `sensor` for Jira Software boards
- HTML5 file drag and drop experience:
  - Make the whole page a HTML5 drop target
  - When entering the target create a `Droppable` and `Draggable` for the dragged item
  - Use a custom sensor to start a drag
  - Use a custom sensor to move the item around in response to user input
- Undo dragging sensor

**Comment by alexreardon** [comment]
This shipped in 12.0! #1487

**Comment by knyzorg** [comment]
@alexreardon 

> ![programatic-dragging 2019-05-07 21_44_54](https://user-images.githubusercontent.com/2182637/57420906-f5e4bc00-724c-11e9-805c-afadfdefb5c4.gif)

Is this implemented in Cypress? I came here looking for how to implement this in a normal context, but this discussion seems to have shifted entirely into Cypress testing, while I just wanted functionality similar to your gif.

Is there code you can show for it?
